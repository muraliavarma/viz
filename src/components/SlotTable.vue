<script>
import SlotTableColumn from './SlotTableColumn.vue';
import SlotTableColumnGroup from './SlotTableColumnGroup.vue';
import SlotTableRow from './SlotTableRow.vue';
import SlotTableCell from './SlotTableCell.vue';

export default {
  name: 'SlotTable',
  // functional components are stateless, but very performant
  functional: true,
  props: {
    tableClass: {
      type: String,
      default: '',
    },
    rows: {
      type: Array,
      required: true,
    },
  },
  // since this is a functional component, data is available through a context argument
  render(createElement, ctx) {
    if (!ctx.children || ctx.children.length === 0) {
      throw new Error('Data Table component expects at least one child');
    }

    // initialize data
    const { rows } = ctx.props;
    const headers = [];
    const cells = [];
    // get all the table listeners
    let rowClick;
    if (ctx.data.on) {
      ({ rowClick } = ctx.data.on);
    }
    // loop through column slots and fetch info from each of its children
    const columns = ctx.children.filter(child => child.data && child.data.slot === 'column');
    columns.forEach((column, columnIndex) => {
      // pass everything into the column component
      const dataTableColumn = createElement(SlotTableColumn, column.data, column.children);
      // retrieve the header and cell component from the column data
      const { header } = dataTableColumn.data.props;

      // header slot should already be ready to be rendered
      headers.push(header);

      // we can not render the cell right here since we need row info
      // thus we retrieve data from the column and store it for later
      const { data } = dataTableColumn;
      cells.push({
        data,
        columnIndex,
      });
    });

    // loop through column slots and fetch info from each of its children
    const groupHeaders = [];
    const columnGroups = ctx.children.filter(
      child => child.data && child.data.slot === 'column-group',
    );
    columnGroups.forEach((columnGroup) => {
      // pass all the into the column component
      const dataTableColumnGroup = createElement(
        SlotTableColumnGroup, columnGroup.data, columnGroup.children,
      );

      // retrieve the header component from the column
      groupHeaders.push(dataTableColumnGroup);
    });

    // construct the table header
    const headerRows = [];
    if (groupHeaders.length > 0) {
      headerRows.push(createElement(SlotTableRow, {}, groupHeaders));
    }
    headerRows.push(createElement(SlotTableRow, {}, headers));
    const thead = createElement('thead', {}, headerRows);

    // loop through each row
    const renderedRows = [];

    rows.forEach((row, rowIndex) => {
      const renderedRow = [];
      // the cell slot might need row info before it can be rendered
      cells.forEach(({ data, columnIndex }) => {
        // pass in the cell and row data for rendering
        const renderedCell = createElement(SlotTableCell, {
          props: {
            data,
            row,
            rowIndex,
            columnIndex,
          },
        });
        // push the cell into the row
        renderedRow.push(renderedCell);
      });

      // pass along the relevant event listener to row
      let rowOn;
      if (rowClick) {
        rowOn = {
          click: () => rowClick(rowIndex),
        };
      }
      // push the row into the list of rows
      renderedRows.push(createElement(SlotTableRow, {
        on: rowOn,
      }, renderedRow));
    });
    // save the list of rows in the tbody
    const tbody = createElement('tbody', {}, renderedRows);

    // finally render out the table with thead, tbody and styles
    const table = createElement('table', {
      attrs: {
        class: ctx.props.tableClass,
      },
    }, [thead, tbody]);

    return table;
  },
};

</script>

<style lang="css">

.sticky-left {
  position: sticky;
  left: 0;
}

.sticky-right {
  position: sticky;
  right: 0;
}

</style>
