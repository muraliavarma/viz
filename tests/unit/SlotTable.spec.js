import { mount } from '@vue/test-utils';
import SlotTable from '@/components/SlotTable.vue';

describe('SlotTable.vue', () => {
  const column = {
    data: {
      slot: 'column',
    },
    children: [],
  };

  it('throws error if there are no children', () => {
    // hide the error logs
    jest.spyOn(global.console, 'error')
      .mockImplementationOnce(() => {})
      .mockImplementationOnce(() => {})
      .mockImplementationOnce(() => {})
      .mockImplementationOnce(() => {});
    expect(() => mount(SlotTable, {
      context: {
        props: {
          rows: [],
        },
      },
    })).toThrow('Data Table component expects at least one child');
  });

  it('renders correct number of rows', () => {
    const wrapper = mount(SlotTable, {
      context: {
        props: {
          rows: ['foo', 'bar', 'baz'],
        },
        children: [column],
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('accepts table class', () => {
    const wrapper = mount(SlotTable, {
      context: {
        props: {
          rows: [],
          tableClass: 'foo',
        },
        children: [column],
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders correct number of columns', () => {
    const wrapper = mount(SlotTable, {
      context: {
        props: {
          rows: [],
        },
        children: [column, column],
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders correct number of column groups', () => {
    const columnGroup = {
      data: {
        slot: 'column-group',
      },
    };
    const wrapper = mount(SlotTable, {
      context: {
        props: {
          rows: [],
        },
        children: [column, columnGroup],
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders header and cells', () => {
    const columnWithSlottedChildren = {
      data: {
        slot: 'column',
      },
      children: [{
        data: {
          slot: 'cell',
        },
      }, {
        data: {
          slot: 'header',
        },
      }],
    };

    const wrapper = mount(SlotTable, {
      context: {
        props: {
          rows: ['foo', 'bar'],
        },
        children: [columnWithSlottedChildren],
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('handles row click', () => {
    const rowClick = i => `clicked row ${i}`;
    const wrapper = mount(SlotTable, {
      context: {
        props: {
          rows: ['foo', 'bar'],
        },
        children: [column],
        on: { rowClick },
      },
    });
    const domRows = wrapper.findAll('tbody tr').wrappers;
    expect(domRows[0].vnode.data.on.click()).toBe('clicked row 0');
    expect(domRows[1].vnode.data.on.click()).toBe('clicked row 1');
  });
});
