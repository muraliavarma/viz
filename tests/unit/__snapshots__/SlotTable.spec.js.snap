// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlotTable.vue accepts table class 1`] = `
<table
  class="foo"
>
  <thead>
    <tr>
      <th />
    </tr>
  </thead>
  <tbody />
</table>
`;

exports[`SlotTable.vue renders correct number of column groups 1`] = `
<table
  class=""
>
  <thead>
    <tr>
      <th />
    </tr>
    <tr>
      <th />
    </tr>
  </thead>
  <tbody />
</table>
`;

exports[`SlotTable.vue renders correct number of columns 1`] = `
<table
  class=""
>
  <thead>
    <tr>
      <th />
      <th />
    </tr>
  </thead>
  <tbody />
</table>
`;

exports[`SlotTable.vue renders correct number of rows 1`] = `
<table
  class=""
>
  <thead>
    <tr>
      <th />
    </tr>
  </thead>
  <tbody>
    <tr>
      <td />
    </tr>
    <tr>
      <td />
    </tr>
    <tr>
      <td />
    </tr>
  </tbody>
</table>
`;

exports[`SlotTable.vue renders header and cells 1`] = `
<table
  class=""
>
  <thead>
    <tr>
      <th>
        undefined
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        undefined
      </td>
    </tr>
    <tr>
      <td>
        undefined
      </td>
    </tr>
  </tbody>
</table>
`;
